edb/edgeql/compiler/stmt.py

```diff
@@ -1049,21 +1049,38 @@ def compile_Shape(
         shape: qlast.Shape, *, ctx: context.ContextLevel) -> irast.Set:
     shape_expr = shape.expr or qlutils.ANONYMOUS_SHAPE_EXPR
     with ctx.new() as subctx:
-        subctx.expr_exposed = False
-        expr = setgen.ensure_set(
-            dispatch.compile(shape_expr, ctx=subctx), ctx=subctx)
-    expr_stype = setgen.get_set_type(expr, ctx=ctx)
-    if not isinstance(expr_stype, s_objtypes.ObjectType):
-        raise errors.QueryError(
-            f'shapes cannot be applied to '
-            f'{expr_stype.get_verbosename(ctx.env.schema)}',
-            context=shape.context,
-        )
-    view_type = viewgen.process_view(
-        stype=expr_stype, path_id=expr.path_id,
-        elements=shape.elements, parser_context=shape.context, ctx=ctx)
+        subctx.qlstmt = shape
```

Comment:
Use `astutils.ensure_qlstmt` here (and use the result as a key below for `compiled_stmts`) and you should avoid the need for a `Union` in the dictionaries, since it seems the value is only used as a key for lookups.

Reply:
Good call